v2/apiserver/internal/core/kubernetes/substrate.go

```diff
@@ -75,12 +105,17 @@ func NewSubstrate(
 	queueWriterFactory queue.WriterFactory,
 	config SubstrateConfig,
 ) core.Substrate {
-	return &substrate{
+	s := &substrate{
 		generateNewNamespaceFn: generateNewNamespace,
 		kubeClient:             kubeClient,
 		queueWriterFactory:     queueWriterFactory,
 		config:                 config,
 	}
+	s.createWorkspacePVCFn = s.createWorkspacePVC
+	s.createWorkerPodFn = s.createWorkerPod
+	s.createJobSecretFn = s.createJobSecret
+	s.createJobPodFn = s.createJobPod
```

Comment:
It took me a second to see the `Fn` at the end there.
(no changes required here, just pointing it out)

Reply:
Ya... will warn that I end up using this pattern in a lot of places. "Constructor" initializes everything correctly, but functions are individually overridable for test purposes. It ends up looking a little complicated, but it improves testability so much.